iree/compiler/Dialect/VM/Conversion/UtilToVM/ConvertAlignmentOps.cpp

```diff
@@ -0,0 +1,85 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/IR/UtilTypes.h"
+#include "iree/compiler/Dialect/VM/Conversion/TypeConverter.h"
+#include "iree/compiler/Dialect/VM/Conversion/UtilToVM/ConvertUtilToVM.h"
+#include "iree/compiler/Dialect/VM/IR/VMOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+class UtilAlignOpConversion : public OpConversionPattern<IREE::Util::AlignOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      IREE::Util::AlignOp srcOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    IREE::Util::AlignOpAdaptor srcOperands(operands);
+    auto valueType = srcOperands.value().getType();
+
+    // (value + (alignment - 1)) & ~(alignment - 1)
+    if (valueType.isInteger(32)) {
+      auto oneConstant =
+          rewriter.createOrFold<IREE::VM::ConstI32Op>(srcOp.getLoc(), 1);
+      // (alignment - 1)
+      auto alignmentValue = rewriter.createOrFold<IREE::VM::SubI32Op>(
+          srcOp.getLoc(), rewriter.getI32Type(), srcOperands.alignment(),
+          oneConstant);
+      // value + (alignment - 1)
+      auto valueAddedValue = rewriter.createOrFold<IREE::VM::AddI32Op>(
+          srcOp.getLoc(), rewriter.getI32Type(), srcOperands.value(),
+          alignmentValue);
+      // ~(alignment - 1)
+      auto notAlignmentValue = rewriter.createOrFold<IREE::VM::NotI32Op>(
+          srcOp.getLoc(), rewriter.getI32Type(), alignmentValue);
+
+      rewriter.replaceOpWithNewOp<IREE::VM::AndI32Op>(
+          srcOp, rewriter.getI32Type(), valueAddedValue, notAlignmentValue);
+    } else if (valueType.isInteger(64)) {
+      auto oneConstant =
+          rewriter.createOrFold<IREE::VM::ConstI64Op>(srcOp.getLoc(), 1);
+      // (alignment - 1)
+      auto alignmentValue = rewriter.createOrFold<IREE::VM::SubI64Op>(
+          srcOp.getLoc(), rewriter.getI32Type(), srcOperands.alignment(),
+          oneConstant);
+      // value + (alignment - 1)
+      auto valueAddedValue = rewriter.createOrFold<IREE::VM::AddI64Op>(
+          srcOp.getLoc(), rewriter.getI64Type(), srcOperands.value(),
+          alignmentValue);
+      // ~(alignment - 1)
+      auto notAlignmentValue = rewriter.createOrFold<IREE::VM::NotI64Op>(
+          srcOp.getLoc(), rewriter.getI64Type(), alignmentValue);
+
+      rewriter.replaceOpWithNewOp<IREE::VM::AndI32Op>(
+          srcOp, rewriter.getI32Type(), valueAddedValue, notAlignmentValue);
+    } else {
+      return srcOp.emitError() << "unsupported value type for util.align";
```

Comment:
you can include the actual type in this error, just add something like `for util.align: " << valueType`